subprojects/messaging/src/main/java/org/gradle/internal/serialize/DefaultSerializerRegistry.java

```diff
@@ -49,42 +52,62 @@ public DefaultSerializerRegistry(boolean supportClassHierarchy) {
 
     @Override
     public <T> void register(Class<T> implementationType, Serializer<T> serializer) {
-        serializerMap.put(implementationType, serializer);
+        readWriteLock.writeLock().lock();
+        try {
+            serializerMap.put(implementationType, serializer);
+        } finally {
+            readWriteLock.writeLock().unlock();
+        }
     }
 
     @Override
     public <T> void useJavaSerialization(Class<T> implementationType) {
-        javaSerialization.add(implementationType);
+        readWriteLock.writeLock().lock();
+        try {
+            javaSerialization.add(implementationType);
+        } finally {
+            readWriteLock.writeLock().unlock();
+        }
     }
 
     @Override
     public boolean canSerialize(Class<?> baseType) {
-        for (Class<?> candidate : serializerMap.keySet()) {
-            if (classMatcher.matches(baseType, candidate)) {
-                return true;
+        readWriteLock.readLock().lock();
+        try {
+            for (Class<?> candidate : serializerMap.keySet()) {
+                if (classMatcher.matches(baseType, candidate)) {
+                    return true;
+                }
             }
-        }
-        for (Class<?> candidate : javaSerialization) {
-            if (classMatcher.matches(baseType, candidate)) {
-                return true;
+            for (Class<?> candidate : javaSerialization) {
+                if (classMatcher.matches(baseType, candidate)) {
+                    return true;
+                }
             }
+        } finally {
```

Comment:
With the proposal of two different ways to guard this code, the lock could be limited to the first block only.

Reply:
This has been resolved.